File: src/data-scratch-library/dsl/c19_deep_learning/layer.py
"""
Neural-network layer abstractions: Layer, Linear, Sequential, and
activation layers (Sigmoid, Tanh, Relu, Dropout).
"""

import math
import operator
import random
from typing import Iterable, List

from dsl.c04_linear_algebra.e0401_vectors import dot
from .deep_learning import Tensor, random_tensor, tensor_apply, tensor_combine


class Layer:
    """Base class for all layers."""

    def forward(self, input: Tensor) -> Tensor:
        raise NotImplementedError

    def backward(self, gradient: Tensor) -> Tensor:
        raise NotImplementedError

    def params(self) -> Iterable[Tensor]:
        return ()

    def grads(self) -> Iterable[Tensor]:
        return ()


class Linear(Layer):
    """Fully-connected layer."""

    def __init__(self, input_dim: int, output_dim: int, init: str = "xavier") -> None:
        self.input_dim = input_dim
        self.output_dim = output_dim
        self.w = random_tensor(output_dim, input_dim, init=init)
        self.b = random_tensor(output_dim, init=init)

    def forward(self, input: Tensor) -> Tensor:
        self.input = input
        return [dot(input, self.w[o]) + self.b[o] for o in range(self.output_dim)]

    def backward(self, gradient: Tensor) -> Tensor:
        self.b_grad = gradient
        self.w_grad = [
            [self.input[i] * gradient[o] for i in range(self.input_dim)]
            for o in range(self.output_dim)
        ]
        return [
            sum(self.w[o][i] * gradient[o] for o in range(self.output_dim))
            for i in range(self.input_dim)
        ]

    def params(self) -> Iterable[Tensor]:
        return [self.w, self.b]

    def grads(self) -> Iterable[Tensor]:
        return [self.w_grad, self.b_grad]


class Sequential(Layer):
    """A sequence of layers applied in order."""

    def __init__(self, layers: List[Layer]) -> None:
        self.layers = layers

    def forward(self, inputs: Tensor) -> Tensor:
        output = inputs
        for layer in self.layers:
            output = layer.forward(output)
        return output

    def backward(self, gradient: Tensor) -> Tensor:
        for layer in reversed(self.layers):
            gradient = layer.backward(gradient)
        return gradient

    def params(self) -> Iterable[Tensor]:
        return (p for layer in self.layers for p in layer.params())

    def grads(self) -> Iterable[Tensor]:
        return (g for layer in self.layers for g in layer.grads())


# -- Activation layers ----------------------------------------------------------

def tanh(x: float) -> float:
    """Numerically stable tanh."""
    if x < -100:
        return -1.0
    if x > 100:
        return 1.0
    em2x = math.exp(-2 * x)
    return (1 - em2x) / (1 + em2x)


class Sigmoid(Layer):
    """Element-wise sigmoid activation."""

    def forward(self, input: Tensor) -> Tensor:
        from dsl.c18_neural_networks.neural_networks import sigmoid
        self.sigmoids = tensor_apply(sigmoid, input)
        return self.sigmoids

    def backward(self, gradient: Tensor) -> Tensor:
        return tensor_combine(
            lambda sig, grad: sig * (1 - sig) * grad,
            self.sigmoids,
            gradient,
        )


class Tanh(Layer):
    """Element-wise tanh activation."""

    def forward(self, input: Tensor) -> Tensor:
        self.tanh = tensor_apply(tanh, input)
        return self.tanh

    def backward(self, gradient: Tensor) -> Tensor:
        return tensor_combine(
            lambda t, g: (1 - t ** 2) * g, self.tanh, gradient
        )


class Relu(Layer):
    """Element-wise ReLU activation."""

    def forward(self, input: Tensor) -> Tensor:
        self.input = input
        return tensor_apply(lambda x: max(x, 0), input)

    def backward(self, gradient: Tensor) -> Tensor:
        return tensor_combine(
            lambda x, g: g if x > 0 else 0, self.input, gradient
        )


class Dropout(Layer):
    """Inverted dropout layer."""

    def __init__(self, p: float) -> None:
        self.p = p
        self.train = True

    def forward(self, input: Tensor) -> Tensor:
        if self.train:
            self.mask = tensor_apply(
                lambda _: 0 if random.random() < self.p else 1, input
            )
            return tensor_combine(operator.mul, input, self.mask)
        return tensor_apply(lambda x: x * (1 - self.p), input)

    def backward(self, gradient: Tensor) -> Tensor:
        if self.train:
            return tensor_combine(operator.mul, gradient, self.mask)
        raise RuntimeError("don't call backward when not in train mode")

File: src/data-scratch-node-library/ts-data-scratch/deep_learning/layer.ts
// Layer implementations for deep learning framework
// Ported from Python layer module

import { Tensor, randomTensorWithInit, tensorApply, tensorCombine, Sigmoid as BaseSigmoid } from './index';
import { dot } from '../linear_algebra';

export abstract class Layer {
    abstract forward(input: Tensor): Tensor;
    abstract backward(gradient: Tensor): Tensor;
    
    params(): Tensor[] {
        return [];
    }
    
    grads(): Tensor[] {
        return [];
    }
}

export class Linear extends Layer {
    private inputDim: number;
    private outputDim: number;
    public w: Tensor;
    public b: Tensor;
    private input: Tensor = [];
    private wGrad: number[][] = [];
    private bGrad: number[] = [];

    constructor(inputDim: number, outputDim: number, init: string = 'xavier') {
        super();
        this.inputDim = inputDim;
        this.outputDim = outputDim;
        
        // Initialize weights and biases
        this.w = randomTensorWithInit(outputDim, inputDim, init);
        this.b = randomTensorWithInit(outputDim, init);
    }

    forward(input: Tensor): Tensor {
        this.input = input;
        // Return vector of neuron outputs
        const result: number[] = [];
        for (let o = 0; o < this.outputDim; o++) {
            const weightedSum = dot(input as number[], (this.w as Tensor[])[o] as number[]) + 
                               (this.b as number[])[o];
            result.push(weightedSum);
        }
        return result;
    }

    backward(gradient: Tensor): Tensor {
        // Bias gradients are the same as output gradients
        this.bGrad = gradient as number[];
        
        // Weight gradients
        this.wGrad = [];
        for (let o = 0; o < this.outputDim; o++) {
            const wGradRow: number[] = [];
            for (let i = 0; i < this.inputDim; i++) {
                wGradRow.push((this.input as number[])[i] * (gradient as number[])[o]);
            }
            this.wGrad.push(wGradRow);
        }
        
        // Input gradients
        const inputGrad: number[] = [];
        for (let i = 0; i < this.inputDim; i++) {
            let sum = 0;
            for (let o = 0; o < this.outputDim; o++) {
                sum += ((this.w as Tensor[])[o] as number[])[i] * (gradient as number[])[o];
            }
            inputGrad.push(sum);
        }
        return inputGrad;
    }

    params(): Tensor[] {
        return [this.w, this.b];
    }
    
    grads(): Tensor[] {
        return [this.wGrad, this.bGrad];
    }
}

export class Sequential extends Layer {
    private layers: Layer[];

    constructor(layers: Layer[]) {
        super();
        this.layers = layers;
    }

    forward(input: Tensor): Tensor {
        let output = input;
        for (const layer of this.layers) {
            output = layer.forward(output);
        }
        return output;
    }

    backward(gradient: Tensor): Tensor {
        let grad = gradient;
        for (let i = this.layers.length - 1; i >= 0; i--) {
            grad = this.layers[i].backward(grad);
        }
        return grad;
    }

    params(): Tensor[] {
        const allParams: Tensor[] = [];
        for (const layer of this.layers) {
            allParams.push(...layer.params());
        }
        return allParams;
    }

    grads(): Tensor[] {
        const allGrads: Tensor[] = [];
        for (const layer of this.layers) {
            allGrads.push(...layer.grads());
        }
        return allGrads;
    }
}

export function tanh(x: number): number {
    if (x < -100) return -1;
    if (x > 100) return 1;
    const em2x = Math.exp(-2 * x);
    return (1 - em2x) / (1 + em2x);
}

export class Tanh extends Layer {
    private tanhOutput: Tensor = [];

    forward(input: Tensor): Tensor {
        this.tanhOutput = tensorApply(tanh, input);
        return this.tanhOutput;
    }

    backward(gradient: Tensor): Tensor {
        return tensorCombine(
            (tanhVal, grad) => (1 - tanhVal * tanhVal) * grad,
            this.tanhOutput,
            gradient
        );
    }
}

export class Relu extends Layer {
    private input: Tensor = [];

    forward(input: Tensor): Tensor {
        this.input = input;
        return tensorApply(x => Math.max(x, 0), input);
    }

    backward(gradient: Tensor): Tensor {
        return tensorCombine(
            (x, grad) => x > 0 ? grad : 0,
            this.input,
            gradient
        );
    }
}

export class Dropout extends Layer {
    private p: number;
    private train: boolean = true;
    private mask: Tensor = [];

    constructor(p: number) {
        super();
        this.p = p;
    }

    setTrainingMode(train: boolean): void {
        this.train = train;
    }

    forward(input: Tensor): Tensor {
        if (this.train) {
            // Create mask using the specified probability
            this.mask = tensorApply(() => Math.random() < this.p ? 0 : 1, input);
            // Multiply by mask to dropout inputs
            return tensorCombine((x, m) => x * m, input, this.mask);
        } else {
            // During evaluation, scale down outputs uniformly
            return tensorApply(x => x * (1 - this.p), input);
        }
    }

    backward(gradient: Tensor): Tensor {
        if (this.train) {
            // Only propagate gradients where mask == 1
            return tensorCombine((g, m) => g * m, gradient, this.mask);
        } else {
            throw new Error("don't call backward when not in train mode");
        }
    }
}

// Re-export Sigmoid from index module for convenience
export { BaseSigmoid as Sigmoid };
